### PR TITLE
Latest version: 29 January 2021

### PR DESCRIPTION
**Pull request by alpakpinar**
Hey @AndreasAlbert, before scaling the system up, I wanted to open a PR with the latest version of the code I use. In this PR we have:

- Update to the latest priors with the partial exponential fit, and HT binnings
- Some updates to the plotting (legend labels with the current code didn't work for me for some reason) 
- Some minor updates to time the script and calculate time/event

Let me know if you have any comments or updates you want to make. Thanks!

### CODE REVIEW

File: rebalance.py [hunk 1]
```diff
@@ -335,24 +334,84 @@ def _name_total_gen_htmiss_prior_pdf(self):
...
+        # For the event at hand, do the following:
+        # 1. Get the histogram corresponding to the HT bin, based on GEN-HT of event
+        # 2. Convert it into a RooDataHist and finally a RooHistPDF
+        # 3. Save the RooHistPDF to the workspace
+
+        ht_bin_for_event = self._figure_out_ht_bin(gen_ht.evaluate())
+
+        # TODO: Year implementation?
+        # For the time being, use 2017 histograms
+        for key in prior_input_file.GetListOfKeys():
```
Comment (AndreasAlbert) [inline rebalance.py:360]: lets refactor this code into independent units. Currently, you have a big loop that is intended to pick the right histogram, and then everything you do with the histogram is inside that loop. Let's instead:
1. factor out the finding of the histogram into a separate method
2. factor out the conversion histogram -> PDF into a separate method

File: rebalance.py [hunk 2]
```diff
@@ -335,24 +334,84 @@ def _name_total_gen_htmiss_prior_pdf(self):
     def _name_total_gen_htmiss_prior_slope(self):
         return 'gen_htmiss_prior_slope'
 
+    def _get_gen_htmiss_prior_file(self):
+        rfile = './input/htmiss_prior.root'
+        return r.TFile(rfile)
+
+    def _figure_out_ht_bin(self, gen_ht):
```
Comment (AndreasAlbert) [inline rebalance.py]: This is a naming convention, right? In that case, let's call it ```_name_ht_bin(self, gen_ht)```

File: rebalance.py [hunk 3]
```diff
@@ -335,24 +334,84 @@ def _name_total_gen_htmiss_prior_pdf(self):
...
+            '500_to_700',
+            '700_to_900',
+            '900_to_1300',
+            '1300_to_2000',
+            '2000_to_5000',
+        ]
+        for ht_bin in prior_ht_bins:
+            ht_bin_split = ht_bin.split('_')
+            min_ht, max_ht = int(ht_bin_split[0]), int(ht_bin_split[-1]) 
+
```
Comment (AndreasAlbert) [inline rebalance.py:348]: can we make this a little bit cleaner?
e.g.:
```python
def _name_ht_bin(ht):
    bins = [100, 300, 500, 700, 900, 1300, 2000, 50000]

    for i in range(len(bins)-1):
        lo = bins[i]
        hi = bins[i+1]
        if lo <= ht <= hi:
            return f'{lo:.0f}_to_{hi:.0f}'
    raise RuntimeError(...)
```

File: rebalance.py [hunk 4]
```diff
@@ -335,24 +334,84 @@ def _name_total_gen_htmiss_prior_pdf(self):
...
-            slope_variable
-        )
-        self._wsimp(prior_pdf)
+        '''Build gen HTmiss prior, extracted from the source file.'''
+        # Get the source file for prior distributions
+        prior_input_file = self._get_gen_htmiss_prior_file()
+
+        # Get the (already saved to WS) HTmiss and HT variables
+        gen_ht = self.ws.function(self._name_total_gen_ht_variable())
+        gen_htmiss = self.ws.function(self._name_partial_gen_htmiss_variable(direction='pt')).evaluate()
```
Comment (AndreasAlbert) [inline rebalance.py]: it's confusing that gen_ht is a function object and gen_htmiss is just a float. let's keep that more consistent

Comment (AndreasAlbert) [reply rebalance.py]: Actually, gen_htmiss is not used -> delete

File: rebalance.py [hunk 5]
```diff
@@ -335,24 +334,84 @@ def _name_total_gen_htmiss_prior_pdf(self):
...
+            histname = hist.GetName()
+
+            if '2018' in histname:
+                continue
+            if ht_bin_for_event not in histname:
+                continue
+
+            htmiss_variable = self.ws.function(self._name_partial_gen_htmiss_variable(direction='pt'))
+
+            dummy_htmiss_variable = r.RooRealVar(
```
Comment (AndreasAlbert) [inline rebalance.py]: maybe add a comment to explain why the dummy variable is needed

File: rebalance.py [hunk 6]
```diff
@@ -3,12 +3,11 @@
 r.gSystem.Load('libRooFit')
 import numpy as np
 
-
 @dataclass(frozen=True)
 class Jet():
     pt: float
-    eta: float
     phi: float
+    eta: float
```
Comment (AndreasAlbert) [inline rebalance.py]: let's please keep the ordering of 'pt, eta, phi'. The solution to the problem you found earlier is not to change the constructor, but to just use it with explicit keyword arguments to preclude any chance of mismatches:
```python
myjet = Jet(pt=10., eta=2.1, phi=0.3)
```

**Review by AndreasAlbert — APPROVED** [review]

**Review by AndreasAlbert — CHANGES_REQUESTED** [review]
I accidentally clicked "approve" before.

**Comment by alpakpinar** [comment]
Hi @AndreasAlbert, thanks a lot for the detailed review and comments! I pushed an additional commit to address your comments. Let me know if you want additional changes to this or if something's missing, thank you!

File: rebalance.py [hunk 7]
```diff
@@ -335,25 +334,88 @@ def _name_total_gen_htmiss_prior_pdf(self):
...
+                        prior_pdf_name,
+                        r.RooArgList(htmiss_variable),
+                        r.RooArgList(dummy_htmiss_variable),
+                        datahist
+                    )
+
+        # Linear interpolation
+        prior_pdf.setInterpolationOrder(1)
+
         self._wsimp(prior_pdf)
```
Comment (alpakpinar) [inline rebalance.py:403]: Just as a quick comment about a detail here:

I re-factored the code that does TH1 -> RooHistPdf conversion and saves the RooHistPdf into the workspace all in one function. If I try to return the pdf to the main function, and attempt to save it there, I ran into seg fault issues, so I moved the saving part into the same function as well.

Comment (AndreasAlbert) [reply rebalance.py:403]: Thanks for clarifying. That's fine.
Two things though:
1. the name of the function sounds very general; one might think that one can use it for all kinds of different roohistpdf creation. However, the naming and htmiss variable choice as input to the roohistpdf are pre-set and cannot be changed, so this function can really only be used for one thing: Creating the ht prior pdf. Suggest to rename to '_build_gen_htmiss_prior_roohistpdf'.
2. What is the point of the `th1name` argument? Seems like a) it doesn't matter in the end, bc the roodatahist is thrown away anyway and b) we currently just put in `th1name=th1.GetName()`, so I do not understand why it needs to be a separate argument

File: rebalance.py (hunk 1 above)
Comment (AndreasAlbert) [reply rebalance.py:360]: Thanks for taking steps in the right direction. However, I think we are not quite there, yet.
1. It does not make sense for `_convert_th1_to_roohistpdf_and_save_to_ws` to be inside this loop. We want the function to be called exactly once, which the current code does not guarantee or make obvious.
2. Your `_get_prior_histogram` is too trivial of a wrapper around `key.ReadObj`.

What I was proposing here is this:
```python

def _get_prior_histogram(self,ht):
	gen_ht = self.ws.function(self._name_total_gen_ht_variable())
	ht_bin_for_event = self._name_ht_bin(gen_ht.evaluate())
	# do whatever is needed to find right th1 and return it
	return th1

def _build_gen_htmiss_prior(self):
	(...)
	th1 = self._get_prior_histogram(ht) # Loop goes in here
	self._convert_th1_to_roohistpdf_and_save_to_ws(th1) # THis is not in a loop

```

**Review by AndreasAlbert — CHANGES_REQUESTED** [review]

**Comment by alpakpinar** [comment]
Hey @AndreasAlbert, thanks a lot for the comments. I now further re-factored the code so that most of the work are done inside the helper functions as you suggested. Let me know if these work for you or if you'd like changes to this. Thanks!

File: rebalance.py (hunk 7 above)
Comment (alpakpinar) [reply rebalance.py:403]: Thanks for the suggestions, function name is updated and th1name argument is removed in a1f7933

**Review by AndreasAlbert — APPROVED** [review]

**Comment by AndreasAlbert** [comment]
thanks for the detailed followup.

**Pull request merged**